app/src/main/java/org/groundplatform/android/ui/datacollection/DataCollectionViewModel.kt

```diff
@@ -74,308 +58,303 @@ import timber.log.Timber
+  private val savedStateHandle: SavedStateHandle,
   @ApplicationScope private val externalScope: CoroutineScope,
   @IoDispatcher private val ioDispatcher: CoroutineDispatcher,
-  private val savedStateHandle: SavedStateHandle,
   private val submissionRepository: SubmissionRepository,
+  private val submitDataUseCase: SubmitDataUseCase,
   private val offlineUuidGenerator: OfflineUuidGenerator,
-  locationOfInterestRepository: LocationOfInterestRepository,
-  surveyRepository: SurveyRepository,
+  private val popups: Provider<EphemeralPopups>,
+  private val viewModelFactory: ViewModelFactory,
+  private val dataCollectionInitializer: DataCollectionInitializer,
 ) : AbstractViewModel() {
 
-  private val _uiState: MutableStateFlow<UiState?> = MutableStateFlow(null)
-  val uiState = _uiState.asStateFlow()
+  private val _uiState = MutableStateFlow<DataCollectionUiState>(DataCollectionUiState.Loading)
+  val uiState: StateFlow<DataCollectionUiState> = _uiState
+
+  val loiNameDialogOpen = mutableStateOf(false)
+  private var shouldLoadFromDraft: Boolean = false
 
   private val jobId: String = requireNotNull(savedStateHandle[TASK_JOB_ID_KEY])
   private val loiId: String? = savedStateHandle[TASK_LOI_ID_KEY]
+  private val loiName: String? = savedStateHandle[TASK_LOI_NAME_KEY]
 
-  /** True iff the user is expected to produce a new LOI in the current data collection flow. */
-  private val isAddLoiFlow = loiId == null
+  private val taskDataHandler = TaskDataHandler()
+  private lateinit var taskSequenceHandler: TaskSequenceHandler
+  private val taskViewModels = MutableStateFlow(mutableMapOf<String, AbstractTaskViewModel>())
 
-  private var shouldLoadFromDraft: Boolean = savedStateHandle[TASK_SHOULD_LOAD_FROM_DRAFT] ?: false
-  private var draftDeltas: List<ValueDelta>? = null
+  private val draftLock = Any()
+  @Volatile private var draftCache: List<ValueDelta>? = null
+  @Volatile private var draftMapCache: Map<Pair<String, Task.Type>, TaskData?>? = null
+  @Volatile private var draftsEnabled = true
 
-  private val activeSurvey: Survey = runBlocking {
-    try {
-      withTimeout(SURVEY_LOAD_TIMEOUT_MILLIS) {
-        surveyRepository.activeSurveyFlow.filterNotNull().first()
+  init {
+    viewModelScope.launch {
+      when (
+        val initResult =
+          dataCollectionInitializer.initialize(savedStateHandle, jobId, loiId, loiName)
+      ) {
+        is DataCollectionUiState.Ready ->
+          run {
+            taskSequenceHandler = TaskSequenceHandler(initResult.tasks, taskDataHandler)
+            _uiState.value = initResult
+          }
+        is DataCollectionUiState.Error ->
+          run {
+            Timber.e(initResult.cause, "Initialization failed code=%s", initResult.code)
+            _uiState.value = initResult
+          }
+        DataCollectionUiState.Loading -> Unit
+        is DataCollectionUiState.TaskUpdated -> Unit
+        DataCollectionUiState.TaskSubmitted -> Unit
       }
```

Comment:
Instead of using `when`, can we just use an if block for checking `Ready` state and then pass the state as is to the `_uiState`?